Improvements to Particle class

paulromano opened the pull request:
This PR makes several changes to the `Particle` class:

- Cross section caches are now a part of the `Particle` class rather than separated as threadprivate variables. Given that you would truly need one cache per particle in flight rather than per thread, it makes more sense for it to be part of the `Particle` class. I also renamed the caches to be a little more descriptive (`neutron_xs_`, `photon_xs_`, and `macro_xs_`).
- The `coord_` member was changed to be a `std::vector` that is dynamically sized. This gets rid of the `MAXCOORD` preprocessor definition.
- The secondary bank was turned into a `std::vector` and was also separated from the particle itself (and made threadprivate). For exploration of things like event-based algorithms, having the secondary bank as a separate vector reduces overall memory requirements.

A few other items of cleanup:

- `fission_bank` is treated like a proper `std::vector`, where we actually `emplace_back()` when a fission site is created. This allows us to just use `size()` to check the size. To avoid memory allocations during runtime, we still `reserve()` memory.
- Indentation of `#pragma`s to make @smharper happy
- Got rid of the `n_bank`, `thread_id`, and `n_threads` global variables

Altogether, I'm seeing a small performance improvement with this branch versus develop (probably as a result of lower memory use thanks to getting rid of statically sized `coord_` and secondary bank?). Would love to see if others can confirm this too.

smharper commented on src/tallies/tally_scoring.cpp:
```diff
@@ -317,7 +317,7 @@ score_fission_eout(const Particle* p, int i_tally, int i_score, int score_bin)
 //! is not used for analog tallies.
 
 void
-score_general_ce(const Particle* p, int i_tally, int start_index,
```
Why is it that this and other tally functions no longer use a `const` pointer to the particle?

paulromano replied:
There's a call to `data::nuclides[i_nuclide]->calculate_elastic_xs(Particle&)` which now modifies the particle since it is storing the microscopic xs cache.

smharper commented:
Nice PR!  It's good to see `MAX_COORD` go as well as a slight trimming down on the number of global variables we use.

smharper commented:
With the check on `MAX_SECONDARY` removed, what will happen if we run a fixed-source mode on a problem with keff > 1?  Will OpenMC just continually build up secondaries until it runs out of memory?  If so, we should probably have some logic to try catching or at least warning about that error.

paulromano commented:
Yes, right now it will just continually grow the secondary bank. Let me think about the best way to handle that.

paulromano commented:
@smharper I added in a check for growth on the secondary bank specifically when a fixed source calculation with fission production is being run. The calculation will abort now if the bank reaches a size >= 10000.

A few other updates:
- Got rid of `ERROR_INT` which was only being used for `Particle.surface_`. Now I just use the value 0 to indicate "not on a surface".
- Cleaned up the interface of `distance_to_boundary` a little bit by having it return a struct with boundary information.
- Renamed `simulation::work` -> `simulation::work_per_rank`
- Fixed a few more compiler warnings (found by gcc 8)
